### PR TITLE
Add python bindings for an HNSW index fixture

### PR DESCRIPTION
**Pull request by toregge**
@geirst : please review

### CODE REVIEW

**Review by geirst — APPROVED** [review]
This looks very good 👍 Some comments and suggestions to consider.

**Comment by geirst** [comment]
I think @aressem and / or @baldersheim should have a look at the module, rpm, and python binding changes.

**Comment by geirst** [comment]
@jobergum FYI

File: ann_benchmark/src/tests/ann_benchmark/test_euclidean.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    f.set(0, [1, 2])
+    assert [1, 2] == f.get(0)
+    f.clear(0)
+    assert [0, 0] == f.get(0)
+
+def test_find():
+    f = Fixture()
+    f.set(0, [0, 0])
+    f.set(1, [10, 10])
+    assert f.get(0) == [0, 0]
```
[the remark this answers is not in the document]
Comment (toregge) [reply ann_benchmark/src/tests/ann_benchmark/test_euclidean.py]: Removed.

File: ann_benchmark/src/vespa/ann_benchmark/vespa_ann_benchmark.cpp [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+    os << "tensor<float>(x[" << dim_size << "])";
+    return os.str();
+}
+
+constexpr uint32_t lid_bias = 1; // lid 0 is reserved
+
+}
+
+/*
+ * Wrapper class for a tensor attribute vector containing a nearest
```
[the remark this answers is not in the document]
Comment (toregge) [reply ann_benchmark/src/vespa/ann_benchmark/vespa_ann_benchmark.cpp]: Suggested change applied.

File: ann_benchmark/src/vespa/ann_benchmark/vespa_ann_benchmark.cpp [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+ * neighbor index.
```
[the remark this answers is not in the document]
Comment (toregge) [reply ann_benchmark/src/vespa/ann_benchmark/vespa_ann_benchmark.cpp]: Suggested change applied.

File: ann_benchmark/src/vespa/ann_benchmark/vespa_ann_benchmark.cpp [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+      _tensor_attribute(nullptr),
+      _nearest_neighbor_index(nullptr),
+      _dim_size(0u)
+{
+    Config cfg(BasicType::TENSOR, CollectionType::SINGLE);
+    _tensor_type = ValueType::from_spec(make_tensor_spec(dim_size));
+    assert(_tensor_type.is_dense());
+    assert(_tensor_type.count_indexed_dimensions() == 1u);
+    _dim_size = _tensor_type.dimensions()[0].size;
+    std::cout << "HnswIndex::HnswIndex Dimension size is " << _dim_size << std::endl;
```
[the remark this answers is not in the document]
Comment (toregge) [reply ann_benchmark/src/vespa/ann_benchmark/vespa_ann_benchmark.cpp]: Removed

File: ann_benchmark/src/vespa/ann_benchmark/vespa_ann_benchmark.cpp [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+        return;
+    }
+    if (size_t(lid + lid_bias) < _attribute->getNumDocs()) {
+        _attribute->clearDoc(lid + lid_bias);
+        _attribute->commit();
+    }
+}
+
+TopKResult
+HnswIndex::find_top_k(uint32_t k, const std::vector<float>& value, uint32_t explore_k, double distance_threshold)
```
[the remark this answers is not in the document]
Comment (toregge) [reply ann_benchmark/src/vespa/ann_benchmark/vespa_ann_benchmark.cpp]: Argument has been removed.

File: dist/vespa.spec [hunk 6]
```diff
@@ -437,6 +437,27 @@ Requires: %{name}-base-libs = %{version}-%{release}
...
+%endif
+%if 0%{?fedora}
+Requires: python3
+%endif
+
+%description ann-benchmark
+
+Vespa - The open big data serving engine - ann-benchmark
+
+Python binding for an HNSW index fixture using tensor attribute.
```
[the remark this answers is not in the document]
Comment (toregge) [reply dist/vespa.spec]: Suggested change applied.

File: ann_benchmark/src/vespa/ann_benchmark/setup.py [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+        libdir="lib.%s-%s-%s" % (sys.platform, platform.machine(), distutils.sysconfig.get_python_version())
+        os.system("mkdir -p build/%s" % libdir)
+        os.system("cp -p vespa_ann_benchmark.*.so build/%s" % libdir)
+
+setup(
+  name="vespa_ann_benchmark",
+  version="0.1.0",
+  author="Tor Egge",
+  author_email="Tor.Egge@verizonmedia.com",
+  description="Python binding for an HNSW index fixture using tensor attribute",
```
[the remark this answers is not in the document]
Comment (toregge) [reply ann_benchmark/src/vespa/ann_benchmark/setup.py]: Suggested change applied

**Review by geirst — APPROVED** [review]
👍

File: dist/vespa.spec [hunk 8]
```diff
@@ -437,6 +437,28 @@ Requires: %{name}-base-libs = %{version}-%{release}
 
 Vespa - The open big data serving engine - tools
 
+%package ann-benchmark
```
Comment (aressem) [inline dist/vespa.spec:440]: This seems very specific. Do we expect to add other benchmarks as well ? In this case we should consider just having vespa-benchmarks as a package.

**Pull request merged**